Update dependencies and add notification_store canister

joel-medicala-yral opened the pull request:
closes https://github.com/dolr-ai/product-roadmap/issues/833

- Updated versions for several dependencies including `thiserror`, `syn`, and `darling`.
- Added `notification_store` canister to the project structure.
- Included `notification_store` in the `shared_utils` module.



#### To check the status of the deployment

- check if the platform orchestrator performed the step to upgrade subnet canister with appropriate version: [Platform Orchestrator function](https://dashboard.internetcomputer.org/canister/74zq4-iqaaa-aaaam-ab53a-cai#get_subnet_last_upgrade_status)
- check the status of upgrade for individual canisters in subnet orchestrators and verify the version. Example for one of the subnet orchesrator: [Subnet Orchestrator function](https://dashboard.internetcomputer.org/canister/rimrc-piaaa-aaaao-aaljq-cai#get_index_details_last_upgrade_status)

To test the platform orchestrator, you can use the following command:

`dfx canister call --query 74zq4-iqaaa-aaaam-ab53a-cai get_subnet_last_upgrade_status --network=ic`
 the canister id (74zq4-iqaaa-aaaam-ab53a-cai) is taken from the canister_ids.json file

uwuxyz committed:
Update dependencies and add notification_store canister

- Updated versions for several dependencies including `thiserror`, `syn`, and `darling`.
- Added `notification_store` canister to the project structure.
- Included `notification_store` in the `shared_utils` module.

uwuxyz committed:
Refactor canister definitions and update dependencies

- Removed deprecated canister definitions from `individual_user_template`, `platform_orchestrator`, and `user_index`.
- Updated `Cargo.toml` for `notification_store` to use workspace configuration for `ic-cdk-timers`.
- Cleaned up unused types and services in the removed canisters to streamline the codebase.

uwuxyz committed:
Add notification_store canister and update individual_user_template

- Introduced the `notification_store` canister with its Candid interface and Rust implementation.
- Updated `individual_user_template` with new types and service definitions to enhance functionality.
- Added export for Candid interface in `notification_store` to ensure compatibility with other canisters.

uwuxyz committed:
Implement authorization checks in notification_store canister

- Updated `add_notification` and `mark_notification_as_read` functions to include authorization checks, returning an error if the caller does not match the user.
- Introduced `NotificationStoreError` enum to handle potential errors in the notification store operations.

ravi-sawlani-yral commented on src/lib/shared_utils/src/canister_specific/notification_store/types/notification.rs:
```diff
@@ -0,0 +1,49 @@
+
+#[derive(Clone, Serialize, Deserialize, CandidType)]
+pub struct VideoUploadPayload {
+    video_id: u64,
```
is this video_uid? if so please rename it to avoid confusion.

ravi-sawlani-yral replied:
serde_alias

uwuxyz committed:
Refactor notification_store canister and update dependencies

- Removed unused dependencies and types from the `notification_store` canister.
- Updated `Cargo.lock` and `Cargo.toml` to reflect changes in dependencies, including the removal of `serde_json` and `futures`.
- Cleaned up code formatting and organization for improved readability and maintainability.

uwuxyz committed:
Update notification_store to use SystemTime for created_at field

- Replaced the use of `ic_cdk::api::time()` with `get_current_system_time_from_ic()` in the `add_notification` and `init` functions for consistency in time handling.
- Changed the type of `created_at` in `NotificationData` from `u64` to `SystemTime` to better represent the timestamp format.

uwuxyz committed:
Revert "Refactor notification_store canister and update dependencies"

This reverts commit 2ffff17d7144979d58574e4a5f34565b831c1b64.

uwuxyz committed:
Revert "Update notification_store to use SystemTime for created_at field"

This reverts commit 376d0e5647ccca1a04fb5606f3d121adb5c9d884.

uwuxyz committed:
Refactor notification_store canister and update time handling

- Removed unused dependencies `serde_json` and `futures` from `Cargo.toml` and `Cargo.lock`.
- Updated the `created_at` field in `NotificationData` to use `SystemTime` for better timestamp representation.
- Replaced `ic_cdk::api::time()` with `get_current_system_time_from_ic()` in relevant functions for consistent time handling.
- Added a `post_upgrade` function to prune notifications older than 30 days using the new time handling method.

uwuxyz committed:
Enhance notification handling in notification_store canister

- Updated the `add_notification` function to append new notifications to the existing list for a user, ensuring notifications are not overwritten.
- Modified the `Notification` struct to derive `Default`, allowing for easier initialization of notification instances.

uwuxyz committed:
Refactor notification_store canister to improve notification handling

- Updated the `add_notification`, `mark_notification_as_read`, and `get_notifications` functions to remove user parameter, using the caller directly for authorization.
- Changed the `created_at` field in `NotificationData` to `SystemTime` for better timestamp representation.
- Introduced `NotificationStoreError` and `Result` types for improved error handling in notification operations.

ravi-sawlani-yral reviewed (commented):
Looks good should we add a basic integration test?

uwuxyz committed:
Enhance notification handling and canister integration

- Updated `mark_notification_as_read` to use `unwrap_or_default()` for safer notification retrieval.
- Added `PartialEq` and `Debug` derives to `LikedPayload`, `VideoUploadPayload`, and `NotificationType` for improved comparison and debugging.
- Introduced `CanisterIdNotificationStore` to `KnownPrincipalType` for better canister management in the environment setup.
- Integrated notification store canister creation and installation in the test environment setup.

uwuxyz committed:
Update notification_store canister path to use gzipped WASM file

- Changed the path for the `notification_store` canister from `.wasm` to `.wasm.gz` in the test environment setup for improved loading efficiency.

uwuxyz committed:
Add notification_store canister to deployment scripts

- Included the `notification_store` canister in the installation and upgrade scripts for both local and IC deployments.
- Updated the build process to ensure the `notification_store` is built and gzipped alongside other canisters.

uwuxyz committed:
Add notification_store canister to CI workflow and test setup

- Included the `notification_store` canister in the GitHub Actions workflow for creation and building.
- Updated the test environment setup to reference the gzipped version of the `notification_store` canister for improved loading efficiency.

uwuxyz committed:
Enhance test environment setup for Pocket IC

- Added multiple calls to `pocket_ic.tick()` in the `get_new_pocket_ic_env` function to simulate time progression.
- Ensured that the environment setup properly reflects the expected state after multiple ticks for improved testing accuracy.

uwuxyz committed:
Update cycle addition in Pocket IC test environment setup

- Increased the cycle amounts added to `platform_canister_id` and `notification_store_canister_id` in the `get_new_pocket_ic_env` function to enhance testing scenarios.
- Adjusted cycle values to ensure the environment reflects a more realistic state for testing purposes.

uwuxyz committed:
Add time progression in download snapshot test

- Introduced a loop to call `pic.tick()` 30 times in the `download_snapshot_test` function to simulate time advancement.
- Enhanced the test setup to better reflect the expected state of the environment after multiple ticks for improved testing accuracy.

uwuxyz committed:
Add cycles to notification store canister in test environment setup

- Added a call to `pocket_ic.add_cycles()` for the `notification_store_canister_id` to ensure it has sufficient cycles for testing.
- Removed redundant cycle addition and tick calls to streamline the environment setup process.

uwuxyz committed:
Update get_notifications query in CRUD test to include pagination parameters

- Modified the `get_notifications` query call in the `test_crud` function to accept pagination parameters (10, 0) for improved testing of notification retrieval functionality.

uwuxyz committed:
10, 0

uwuxyz committed:
Update get_notifications query in guard test to include pagination parameters

- Modified the `get_notifications` query calls in the `test_guard` function to accept pagination parameters (10, 0) for improved testing of notification retrieval functionality.

uwuxyz committed:
Update get_notifications query in CRUD and guard tests to include pagination parameters

- Modified the `get_notifications` query calls in both the `test_crud` and `test_guard` functions to accept pagination parameters (10, 0) for improved testing of notification retrieval functionality.

uwuxyz committed:
Refactor get_notifications query in CRUD and guard tests to use Encode macro

- Updated the `get_notifications` query calls in both `test_crud` and `test_guard` functions to utilize the `Encode!` macro for improved readability and consistency in encoding pagination parameters.

uwuxyz committed:
Enhance notification_store canister installation and upgrade process

- Updated the installation commands for the `notification_store` canister to include version arguments for both upgrade and local deployment scripts.
- Refactored the `notification_store` canister's data structure to include a version field, improving version management.
- Modified the notification handling logic to utilize a more structured approach with `CanisterData`, enhancing readability and maintainability.
- Added a new module for initialization arguments to streamline canister setup.

uwuxyz committed:
Refactor get_notifications query calls in CRUD and guard tests to use reference parameters

- Updated the `get_notifications` query calls in both `test_crud` and `test_guard` functions to use reference parameters in the `Encode!` macro, enhancing consistency and clarity in the encoding process.

uwuxyz committed:
Refactor get_notifications query calls in CRUD and guard tests to use u64 parameters

- Updated the `get_notifications` query calls in both `test_crud` and `test_guard` functions to use `u64` for pagination parameters in the `Encode!` macro, ensuring type consistency and improving clarity in the encoding process.

uwuxyz committed:
Refactor get_notifications query calls in CRUD and guard tests to use u64 parameters

- Updated the `get_notifications` query calls in both `test_crud` and `test_guard` functions to use `u64` for pagination parameters in the `Encode!` macro, ensuring type consistency and improving clarity in the encoding process.

uwuxyz committed:
Refactor mark_notification_as_read update call in CRUD test to use u64 parameter

uwuxyz committed:
Fix parameter encoding in mark_notification_as_read call in CRUD test

- Updated the `mark_notification_as_read` call in the `test_crud` function to use a direct `u64` value instead of wrapping it in an additional expression, enhancing clarity and consistency in the encoding process.

uwuxyz committed:
Update test_increment_notification_id to validate notification retrieval

- Modified the `test_increment_notification_id` function to query notifications instead of the next notification ID, enhancing the test to assert the correct retrieval of notifications and their IDs.

uwuxyz committed:
update did

uwuxyz committed:
Refactor notification_store canister structure and remove unused code

- Introduced a new `data_model` module and refactored the `CanisterData` structure for improved organization.
- Removed unused memory management code and notification handling functions to streamline the notification store canister.
- Updated imports in lifecycle files to reflect the new module structure.

uwuxyz committed:
Remove unused NEXT_ID reference and simplify notification ID assignment in add_notification function

ravi-sawlani-yral commented on src/canister/notification_store/src/data_model/mod.rs:
```diff
@@ -0,0 +1,13 @@
+
+#[derive(CandidType, Serialize, Deserialize, Default)]
+pub struct CanisterData {
+    pub notifications: BTreeMap<Principal, Notification>,
```
This should be stable BTreemap, right?

ravi-sawlani-yral commented on src/lib/integration_tests/tests/notification_store/guard.rs:
```diff
@@ -0,0 +1,44 @@
+use test_utils::setup::{env::pocket_ic_env::get_new_pocket_ic_env, test_constants::{get_mock_user_alice_principal_id, get_mock_user_charlie_principal_id}};
+
+#[test]
+fn test_guard() {
```
```suggestion
fn test_notification_authorization() {
```

ravi-sawlani-yral commented on src/lib/shared_utils/src/canister_specific/notification_store/types/notification.rs:
```diff
@@ -0,0 +1,50 @@
+#[derive(Clone, Serialize, Deserialize, CandidType, PartialEq, Debug)]
+pub struct VideoUploadPayload {
+    #[serde(alias = "video_uid")]
+    pub video_id: u64,
```
the field should be `video_uid` and the alias could be video_id so that in code we don't get confused.

uwuxyz committed:
Refactor notification handling in notification_store canister

- Updated the `add_notification` and `mark_notification_as_read` functions to improve clarity and efficiency in notification management.
- Simplified the notification data structure and adjusted memory management for better performance.
- Enhanced the `CanisterData` structure to utilize `StableBTreeMap` for notifications, ensuring more efficient data handling.
- Updated tests to reflect changes in notification ID handling and parameter encoding.

uwuxyz committed:
Update pruning timer interval in notification_store canister

- Renamed the constant for the pruning interval to improve clarity.
- Updated the timer setting function to use the new constant for better readability and maintainability.

joel-medicala-yral committed:
Merge branch 'main' into joel/notification_store

uwuxyz committed:
Refactor notification store integration tests to use new environment setup

- Updated tests in `crud.rs` and `guard.rs` to utilize `get_new_pocket_ic_env_with_service_canisters_provisioned` for improved clarity and consistency.
- Removed references to `KnownPrincipalType::CanisterIdNotificationStore` as it is no longer needed.
- Adjusted the structure of the `ServiceCanisters` to include `notification_store_canister_id` for better management of service canisters.

uwuxyz committed:
Update notification store integration test to use known principals for canister ID retrieval

- Changed the assignment of `notification_store_canister_id` to utilize `known_principals` instead of `service_canisters`, improving clarity and consistency in the test setup.

uwuxyz committed:
Refactor test_increment_notification_id to use updated environment setup

- Changed the test to utilize `get_new_pocket_ic_env_with_service_canisters_provisioned` for improved clarity.
- Simplified the retrieval of `notification_store_canister_id` by directly accessing it from `known_principals`.

uwuxyz committed:
Add cycles to notification store canister in environment setup

- Updated the `get_new_pocket_ic_env_with_service_canisters_provisioned` function to provision cycles for the `notification_store_canister`, ensuring it is adequately funded for testing scenarios.

Commit:
update candid generator and remove unused coded
